### PR TITLE
SparkCore Init & ReadFile implementation

### PR DESCRIPTION
**Pull request by rabbitonweb**
This PR implements `ReadFile` algebra for SparkCore connector. Details in https://github.com/quasar-analytics/quasar/issues/1253

### CODE REVIEW

**Comment by rabbitonweb** [comment]
PR ready to review!

There are still few things to be improved (below). Those will be added asap, however the "main" functionality is ready and will not suffer from those changes. 


Missing bits are:
1. write a test ` pass test open file -> read chunk until empty vector -> close == file content`, could anyone willing to PP on it?
2. tests should call `sc.stop()` regardless of whether they've passed or fail (there can be only one SC in the single JVM)
3. `SparkContextIO` should be just type alias to `ReaderT`
4. `ReadFile.sparkCursorL` should get  better name (still thinking on the name :) )

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/readfile.scala [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+      Lens.lensu[SparkCursor, Option[RDD[Data]]] (
+        (spCur, v) => spCur.copy(rdd = v),
+        _.rdd
+      )
+
+    def maybeRddL(h: ReadHandle): ReadState @?> Option[RDD[Data]] =
+      ~sparkCursorL compose PLens.somePLens compose ~cursorL(h)
+
+  }
+
```
Comment (rabbitonweb) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/readfile.scala]: @jedesah mentioned changing signature to `ReadFile ~> Free` (correct me if I'm wrong) but I don't really want to do it on my own, would really benefit from PP here

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/local_readfile.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import quasar.Predef._
+import quasar.fs._
+import org.apache.spark.rdd._
+
+import java.io.File
+import pathy.Path.posixCodec
+
+import scalaz._
+import Scalaz._
+
```
Comment (rabbitonweb) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/local_readfile.scala]: `spark-core` is a foundation (a base) to build future Spark connectors. First production ready connector will be `spark-hdfs` that will work on Hadoop distributed file system. However for now, to "prove" correctness of `spark-core` I implement also sort of "dummy" connectr  called spark-local.

the idea of spark-local is to run on a single-node spark cluster where all files are on the same file system (not a distributed file system)

Comment (wemrysi) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/local_readfile.scala]: Does spark have different capabilities or limitations based on the backend you use? Just trying to understand what will be different about `spark-hdfs` vs, say, `spark-cassandra`.

File: project/Dependencies.scala [hunk 3]
```diff
@@ -15,7 +15,8 @@ object Dependencies {
   private val scalazVersion   = "7.2.3"
   private val scodecVersion   = "1.0.10"
   private val slcVersion      = "0.4"
-
+  private val sparkVersion    = "1.6.2"
```
Comment (jedesah) [inline project/Dependencies.scala]: I don't think we need this anymore.

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/local_readfile.scala (hunk 2 above)
Comment (rabbitonweb) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/local_readfile.scala]: @wemrysi let me answer with an example:

for ReadFile majority of code will be common for both `spark-hdfs` and `spark-cassandra`, the only difference is how we going to construct the `RDD[String` and those will be implemented by `spark-hdfs` and `spark-cassandra` separately

for WriteFile there is really no common code so each of the subprojects will have to provide its own interpretation of what that is

**Comment by rabbitonweb** [comment]
@jr: rdy right?

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/local_readfile.scala [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import java.io.File
+
+import org.apache.spark.SparkContext
+import org.apache.spark.rdd._
+import pathy.Path.posixCodec
+import scalaz._
+import scalaz.concurrent.Task
+
+object local_readfile {
```
Comment (wemrysi) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/local_readfile.scala]: We've tended to avoid snake case thus far, maybe just name the object `local`?

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/readfile.scala [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  rddFrom: AFile => Free[S, RDD[String]],
+  fileExists: AFile => Free[S, Boolean]
+)
+
+object readfile {
+
+  import ReadFile.ReadHandle
+
+  def interpret[S[_]](input: Input[S])(implicit
+    s0: KeyValueStore[ReadHandle, SparkCursor, ?] :<: S,
```
Comment (wemrysi) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/readfile.scala]: Something we've started doing elsewhere is making aliases for these partially applied effects, i.e.
```scala
type SparkReadCursors[A] = KeyValueStore[ReadHandle, SparkCursor, A]

object SparkReadCursors {
  def Ops[S[_]](implicit S: SparkReadCursors :<: S) =
    KeyValueStore.Ops[ReadHandle, SparkCursor, S]
}
```
which makes them a bit more meaningful, I think.

**Comment by wemrysi** [comment]
A couple of minor comments, but otherwise LGTM :shipit:

File: build.sbt [hunk 6]
```diff
@@ -202,14 +202,16 @@ lazy val core = project
     ScoverageKeys.coverageMinimum := 79,
     ScoverageKeys.coverageFailOnMinimum := true,
     buildInfoKeys := Seq[BuildInfoKey](version, ScoverageKeys.coverageEnabled),
-    buildInfoPackage := "quasar.build")
+    buildInfoPackage := "quasar.build"
+  )
```
Comment (jedesah) [inline build.sbt]: This is an unnecessary change and not the style adopted in the rest of the codebase.

File: build.sbt [hunk 7]
```diff
@@ -240,6 +242,16 @@ lazy val postgresql = project
   .settings(publishSettings: _*)
   .enablePlugins(AutomateHeaderPlugin)
 
+lazy val sparkcore = project
+  .settings(name := "quasar-sparkcore-internal")
+  .dependsOn(core % "test->test;compile->compile")
+  .settings(oneJarSettings: _*)
+  .settings(publishSettings: _*)
+  .settings(libraryDependencies +=
+    "org.apache.spark"  %  "spark-core_2.11"           % "1.6.2")
```
Comment (jedesah) [inline build.sbt]: Perhaps we can use `%%` operator?

File: build.sbt [hunk 8]
```diff
@@ -240,6 +242,16 @@ lazy val postgresql = project
   .settings(publishSettings: _*)
   .enablePlugins(AutomateHeaderPlugin)
 
+lazy val sparkcore = project
+  .settings(name := "quasar-sparkcore-internal")
+  .dependsOn(core % "test->test;compile->compile")
```
Comment (jedesah) [inline build.sbt]: We can use `BothScopes` now.

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/local_readfile.scala (hunk 4 above)
Comment (rabbitonweb) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/local_readfile.scala]: this will probably end up in it's own subproject (will emerge spark-local subproject), can add it to single package for now

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/readfile.scala (hunk 5 above)
Comment (rabbitonweb) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/readfile.scala]: This is what I;ve seen in Noel Welsh blog post when he was introducing monad trnasformes - he used the same trick here

So should I go with type aliasing like that and remove all `?`  ??

File: build.sbt (hunk 6 above)
Comment (rabbitonweb) [inline build.sbt]: Will revert

File: build.sbt (hunk 7 above)
Comment (rabbitonweb) [inline build.sbt]: I will leave it for now as it is as there is one issue I have to talk with John regarding scala version & spark

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/readfile.scala (hunk 5 above)
Comment (wemrysi) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/readfile.scala]: I think it's a good idea, helps to separate semantics from implementation.

File: build.sbt [hunk 9]
```diff
@@ -240,6 +241,16 @@ lazy val postgresql = project
   .settings(publishSettings: _*)
   .enablePlugins(AutomateHeaderPlugin)
 
+lazy val sparkcore = project
+  .settings(name := "quasar-sparkcore-internal")
+  .dependsOn(core % BothScopes)
+  .settings(oneJarSettings: _*)
+  .settings(publishSettings: _*)
+  .settings(libraryDependencies +=
+    "org.apache.spark"  %  "spark-core_2.11"           % "1.6.2")
```
Comment (jedesah) [inline build.sbt]: Still no conclusion on this from your discussion with John? I am not sure I see why we would ever want to use a version that is not compiled with the same of scala as our own. Anyway, I trust you have a good reason for wanting to do it this way.

**Comment by jedesah** [comment]
I think there are a few of @wemrysi's concerns that you did not fully address.

I think it's OK to merge, I expect it to change some more soon so I won't be too picky since this PR has been open for a while now and it would be good to have in `master` for everyone to look at and base work on.

I noticed that some commits have compile errors in them, so I won't merge like we talked about in the retrospective meeting today. Instead, I will mark it as ready and let you rebase and merge in order to remove the compile errors.
